### PR TITLE
flamenco, tiles: added notification pipe to replay to support websockets in rpcserver

### PR DESCRIPTION
**Pull request by asiegel-jt**
…ebsockets in rpc server

### CODE REVIEW

File: src/flamenco/runtime/fd_acc_mgr.c [hunk 1]
```diff
@@ -317,7 +317,12 @@ fd_acc_mgr_save_non_tpool( fd_acc_mgr_t *          acc_mgr,
...
     FD_LOG_ERR(( "unable to allocate account value, err %d", err ));
   }
-  return fd_acc_mgr_save( acc_mgr, account );
+  err = fd_acc_mgr_save( acc_mgr, account );
+  fd_acc_saved_fun saved_fun = acc_mgr->saved_fun;
+  if( saved_fun != NULL ) {
+    (*saved_fun)( account->pubkey, txn, acc_mgr->saved_fun_arg );
+  }
+  return err;
 }
```
Comment (lheeger-jump) [inline src/flamenco/runtime/fd_acc_mgr.c:326]: why a callback? this seems like a bad pattern when the replay tile has all of the info already to push these things out

Comment (asiegel-jt) [reply src/flamenco/runtime/fd_acc_mgr.c:326]: i don't think the replay tiles knows which accounts were updated

File: src/app/fdctl/run/tiles/fd_replay.c [hunk 2]
```diff
@@ -277,6 +272,36 @@ after_frag( void *             _ctx,
...
+        }
+      }
+      if( msg ) {
+        fd_mcache_publish( ctx->notif_out_mcache, ctx->notif_out_depth, ctx->notif_out_seq, 0UL, ctx->notif_out_chunk,
+                           sizeof(fd_replay_notif_msg_t), 0UL, 0UL, 0UL );
+        ctx->notif_out_seq   = fd_seq_inc( ctx->notif_out_seq, 1UL );
+        ctx->notif_out_chunk = fd_dcache_compact_next( ctx->notif_out_chunk, sizeof(fd_replay_notif_msg_t),
+                                                       ctx->notif_out_chunk0, ctx->notif_out_wmark );
+      }
+
```
Comment (lheeger-jump) [inline src/app/fdctl/run/tiles/fd_replay.c:304]: We may want to send some slot metadata at some point, but it is fine for now if you do not need it. For example, slot number, blockhash. It is mostly in the bank anyways so you can decode it but just a thought.

**Review by lheeger-jump — DISMISSED** [review]

File: src/app/fdctl/run/tiles/fd_replay.c [hunk 3]
```diff
@@ -277,6 +272,36 @@ after_frag( void *             _ctx,
...
+        if( !fd_funk_key_is_acc( rec->pair.key ) ) continue;
+        if( msg == NULL ) {
+          msg = fd_chunk_to_laddr( ctx->notif_out_mem, ctx->notif_out_chunk );
+          msg->type = FD_REPLAY_SAVED_TYPE;
+          msg->acct_saved.funk_xid = rec->pair.xid[0];
+          msg->acct_saved.acct_id_cnt = 0;
+        }
+        fd_memcpy( msg->acct_saved.acct_id[ msg->acct_saved.acct_id_cnt++ ].uc, rec->pair.key->uc, sizeof(fd_pubkey_t) );
+        if( msg->acct_saved.acct_id_cnt == FD_REPLAY_NOTIF_ACCT_MAX ) {
+          fd_mcache_publish( ctx->notif_out_mcache, ctx->notif_out_depth, ctx->notif_out_seq, 0UL, ctx->notif_out_chunk,
```
Comment (arjain4) [inline src/app/fdctl/run/tiles/fd_replay.c:289]: nit: you can add tsorig/tspub here for completeness

**Review by arjain4 — APPROVED** [review]

File: src/app/fdctl/run/tiles/fd_replay_notif.h [hunk 4]
```diff
@@ -0,0 +1,21 @@
+/* Data structure which is passed through replay_notif link */
```
Comment (arjain4) [inline src/app/fdctl/run/tiles/fd_replay_notif.h:1]: nit: add header def

File: src/app/fdctl/run/tiles/fd_replay.c (hunk 2 above)
Comment (asiegel-jt) [reply src/app/fdctl/run/tiles/fd_replay.c:304]: the slot number is in the xid

**Review by lidatong — APPROVED** [review]

**Pull request merged**